ReadFileInto skips a single leading UTF8 BOM sequence if it exists.

wanglei-ok opened the pull request:
For compatibility with files created on Windows, ReadFileInto skips a single leading UTF8 BOM sequence if it exists.

- Add ReadFileInto documentation 
- Add test data testdata/notepad.ini
- Add testcase  TestReadFileIntoNotepad

[Related pull request#13](https://github.com/go-gcfg/gcfg/pull/13)

wanglei-ok committed:
ReadFileInto skips a single leading UTF8 BOM sequence if it exists.

speter commented on read.go:
```diff
@@ -231,6 +234,15 @@ func ReadFileInto(config interface{}, filename string) error {
 	}
+
+	//Skips a single leading UTF8 BOM sequence if it exists.
+	if len(src) > 3 {
```
1. `>` should be `>=`
2. make this a helper function, say skipLeadingUtf8Bom([]byte) []byte
3. the end-to-end test is fine, but please add table-driven unit tests for this new helper function:
   - 0 bytes input
   - 3 bytes input (BOM only)
   - 3 bytes input (comment only, without BOM)
   - normal input with BOM
   - normal input without BOM

speter replied:
Starting to look good. A couple more requests.

wanglei-ok committed:
1. > to >=
2. Make helper function, say skipLeadingUtf8Bom([]byte) []byte
3. Add table-driven unit tests for skipLeadingUtf8Bom

wanglei-ok commented:
Thanks for your guidance. Please take another look.

speter commented on read.go:
```diff
@@ -231,7 +234,23 @@ func ReadFileInto(config interface{}, filename string) error {
 }
+
+func skipLeadingUtf8Bom(src []byte) []byte{
+	if len(src) >= 3 {
```
1. create global `var utf8Bom = []byte("\ufeff")`
2. use `len(utf8Bom)` instead of `3`
3. get rid of variable `bom`, join the two `if`s into one, and use `bytes.Equal` with utf8Bom for comparison

speter commented on read_test.go:
```diff
@@ -404,3 +416,28 @@ func TestPanics(t *testing.T) {
+	out []byte
+}{
+	{"0 bytes input",[]byte{}, []byte{}},
+	{"3 bytes input (BOM only)",[]byte{0xEF,0xBB,0xBF}, []byte{}},
```
use `[]byte("\ufeff")` for consistency. same 2 lines below.

speter reviewed (changes requested):
Overall looking very good. Now we are fully working, tested, documented. Just some more minor cosmetic changes.

Also, make sure to run gofmt before commit. You should always do this when contributing Go code.

wanglei-ok committed:
Make the code look more beautiful.

wanglei-ok commented:
These changes make them better.

speter commented:
Thanks!
